bentoml/yatai/yatai_service_impl.py

```diff
@@ -578,6 +584,79 @@ def ContainerizeBento(self, request, context=None):
+                                else:
+                                    raise BentoMLException(
+                                        f"Incoming stream request doesn't match "
+                                        f"with initial request info "
+                                        f"{bento_name}:{bento_version} - "
+                                        f"{request.bento_name}:"
+                                        f"{request.bento_version}"
+                                    )
+                        bento_id = f"{bento_name}_{bento_version}"
+                        with lock(self.db, [(bento_id, LockType.WRITE)]) as (sess, _):
+                            bento_pb = self.db.metadata_store.get(
+                                sess, bento_name, bento_version
+                            )
+                            file.seek(0)
+                            if bento_pb:
+                                with tarfile.open(fileobj=file, mode='r') as tar:
+                                    tar.extractall(path=bento_pb.uri.uri)
+                                result_status = Status.OK()
+                            else:
+                                result_status = Status.NOT_FOUND(
+                                    "BentoService `{}:{}` is not found".format(
+                                        bento_name, bento_version
+                                    )
+                                )
+                    return UploadBentoResponse(status=result_status)
+            except BentoMLException as e:
+                logger.error("RPC ERROR UploadBento: %s", e)
+                return UploadBentoResponse(status=e.status_proto)
+            except Exception as e:  # pylint: disable=broad-except
+                logger.error("RPC ERROR UploadBento: %s", e)
+                return UploadBentoResponse(status=Status.INTERNAL())
+
+        def DownloadBento(self, request, context=None):
+            bento_id = f"{request.bento_name}_{request.bento_version}"
+            with lock(self.db, [(bento_id, LockType.READ)]) as (sess, _):
+                try:
+                    bento_pb = self.db.metadata_store.get(
+                        sess, request.bento_name, request.bento_version
+                    )
+                    responses_generator = BentoBundleStreamRequestsOrResponses(
+                        bento_name=request.bento_name,
+                        bento_version=request.bento_version,
+                        directory_path=bento_pb.uri.uri,
+                        is_request=False,
+                    )
+                    for response in responses_generator:
+                        yield response
+                except BentoMLException as e:
+                    responses_generator.clear()
+                    logger.error("RPC ERROR DownloadBento: %s", e)
+                    return DownloadBentoResponse(status=e.status_proto)
```

Comment:
We may be able to use a simpler pattern by throwing the exception back into the generator with `response_generator.throw(e)`. Then inside the generator, we can handle the exception and close the files in a `finally` block. Let's quickly chat over Zoom if my explanation isn't clear.

Reply:
I understand your explanation.  Let's still have a quick zoom call with where the responsibility ends for the generator.

Reply:
Discussed offline, will invoke `close()` in `finally`.